DREAM/Optimization/tsgOptimizationUtils.hpp

```diff
@@ -0,0 +1,87 @@
+ * IMPLIED. THERE ARE NO EXPRESS OR IMPLIED WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE, OR THAT THE USE OF
+ * THE SOFTWARE WILL NOT INFRINGE ANY PATENT, COPYRIGHT, TRADEMARK, OR OTHER PROPRIETARY RIGHTS, OR THAT THE SOFTWARE WILL
+ * ACCOMPLISH THE INTENDED RESULTS OR THAT THE SOFTWARE OR ITS USE WILL NOT RESULT IN INJURY OR DAMAGE. THE USER ASSUMES
+ * RESPONSIBILITY FOR ALL LIABILITIES, PENALTIES, FINES, CLAIMS, CAUSES OF ACTION, AND COSTS AND EXPENSES, CAUSED BY, RESULTING
+ * FROM OR ARISING OUT OF, IN WHOLE OR IN PART THE USE, STORAGE OR DISPOSAL OF THE SOFTWARE.
+ */
+
+#ifndef __TASMANIAN_OPTIM_ENUMERATES_HPP
+#define __TASMANIAN_OPTIM_ENUMERATES_HPP
+
+#include "TasmanianDREAM.hpp"
+#include <assert.h>
+
+namespace TasOptimization {
+
+inline void checkVarSize(std::string var_name, int var_size, int exp_size) {
+    if (var_size != exp_size) {
+        throw std::runtime_error("Size of " + var_name + " (" + std::to_string(var_size) + ") is not equal to its expected value (" +
+                                 std::to_string(exp_size) + ")");
+    }
+}
+
+using ObjectiveFunctionSingle = std::function<double(const std::vector<double> &x)>;
+
+using ObjectiveFunction = std::function<void(const std::vector<double> &x_batch, std::vector<double> &fval_batch)>;
+
+inline ObjectiveFunction makeObjectiveFunction(int ndim, ObjectiveFunctionSingle f_single) {
+    return [=](const std::vector<double> &x_values, std::vector<double> &fval_values)->void {
+        int num_points = x_values.size() / ndim;
+        std::vector<double> x(ndim);
+        for (int i=0; i<num_points; i++) {
+            std::copy(x_values.begin() + i * ndim, x_values.begin() + (i + 1) * ndim, x.begin());
+            fval_values[i] = f_single(x);
+        }
+    };
+}
+
+// If inside_batch[i] is true, then fval_batch[i] is the function value of a point in the domain; otherwise, if inside_batch[i]
+// is false, then fval_batch[i] is std::numeric_limits<double>::max().
+using ObjectiveFunctionConstrained = std::function<void(const std::vector<double> &x_batch, std::vector<double> &fval_batch,
+                                                        std::vector<bool> &inside_batch)>;
+
+inline ObjectiveFunctionConstrained makeObjectiveFunctionConstrained(int ndim, ObjectiveFunction f, TasDREAM::DreamDomain inside) {
+    return [=](const std::vector<double> &x_batch, std::vector<double> &fval_batch, std::vector<bool> &inside_batch)->void {
+        std::vector<double> candidate(ndim);
+        std::vector<double> single_val(1);
+        for (size_t i=0; i<fval_batch.size(); i++) {
+            std::copy(x_batch.begin() + i * ndim, x_batch.begin() + (i + 1) * ndim, candidate.begin());
+            inside_batch[i] = inside(candidate);
+            fval_batch[i] = std::numeric_limits<double>::max();
+            if (inside_batch[i]) {
+                f(candidate, single_val);
```

Comment:
The logic is correct, but this violates the desire for batch calls, since you're making calls on a point-by-point basis.

Do one sweep and resolve the `inside` call for all points. Then make a second sweep where you copy all good points into a single vector and make a single call to `f()`. Copy the result back to `fval_batch` in the correct order. This incurs an extra copy, but it is still better to make batch calls with one copy as opposed to point-by-point calls.

Reply:
Changed.